transcendence-app/src/auth/auth.controller.ts

```diff
@@ -190,4 +216,108 @@ export class AuthController {
+    return this.authService.pipeQrCodeStream(response, otpAuthUrl);
+  }
+
+  @Post('2fa')
+  @HttpCode(HttpStatus.NO_CONTENT)
+  @UseGuards(TwoFactorAuthenticatedGuard)
+  @ApiNoContentResponse({ description: 'Enable 2FA' })
+  @ApiBadRequestResponse({ description: 'Wrong authentication code' })
+  @ApiForbiddenResponse({ description: 'Forbidden' })
+  @ApiServiceUnavailableResponse({ description: 'Service Unavailable' })
+  async enableTwoFactorAuthentication(
+    @Res({ passthrough: true }) response: Response,
+    @GetUser('id') userId: string,
+    @Body() { twoFactorAuthenticationCode }: TwoFactorAuthenticationCodeDto,
+  ) {
+    const isValidCode =
+      await this.authService.isTwoFactorAuthenticationCodeValid(
+        twoFactorAuthenticationCode,
+        userId,
+      );
+
+    if (!isValidCode) {
+      throw new BadRequestException('Wrong authentication code');
+    }
+    const updatedUser = await this.userService.enableTwoFactorAuthentication(
+      userId,
+    );
+    if (!updatedUser) {
+      throw new ServiceUnavailableException();
+    }
+    response.cookie(
+      IS_TWO_FACTOR_AUTHENTICATED_COOKIE_NAME,
+      IS_TWO_FACTOR_AUTHENTICATED_COOKIE_VALUE,
+      { signed: true },
+    );
+  }
+
+  @Delete('2fa')
+  @HttpCode(HttpStatus.NO_CONTENT)
+  @UseGuards(TwoFactorAuthenticatedGuard)
+  @ApiNoContentResponse({ description: 'Disable 2FA' })
+  @ApiForbiddenResponse({ description: 'Forbidden' })
+  @ApiNotFoundResponse({ description: 'Not Found' })
+  @ApiServiceUnavailableResponse({ description: 'Service Unavailable' })
+  async disableTwoFactorAuthentication(
+    @Res({ passthrough: true }) response: Response,
+    @GetUser() user: User,
+  ) {
+    if (!user.isTwoFactorAuthenticationEnabled) {
+      throw new NotFoundException();
+    }
+    const updatedUser = await this.userService.disableTwoFactorAuthentication(
+      user.id,
+    );
+    if (!updatedUser) {
+      throw new ServiceUnavailableException();
+    }
+    response.clearCookie(IS_TWO_FACTOR_AUTHENTICATED_COOKIE_NAME, {
+      signed: true,
+    });
+  }
+
+  @Post('2fa/validate')
+  @HttpCode(HttpStatus.OK)
+  @UseGuards(AuthenticatedGuard)
+  @ApiOkResponse({ description: 'Validate 2FA' })
+  @ApiBadRequestResponse({ description: 'Wrong authentication code' })
+  @ApiForbiddenResponse({ description: 'Forbidden' })
+  validateTwoFactorAuthentication(
+    @Res({ passthrough: true }) response: Response,
+    @GetUser() user: User,
+    @Body() { twoFactorAuthenticationCode }: TwoFactorAuthenticationCodeDto,
+  ) {
+    const isValidCode = this.authService.isTwoFactorAuthenticationCodeValid(
+      twoFactorAuthenticationCode,
+      user.id,
+    );
+    if (!isValidCode) {
+      throw new BadRequestException('Wrong authentication code');
+    }
+    response.cookie(
+      IS_TWO_FACTOR_AUTHENTICATED_COOKIE_NAME,
+      IS_TWO_FACTOR_AUTHENTICATED_COOKIE_VALUE,
+      { signed: true },
```

Comment:
Why do you need this payload for? If I understood well, cookie-parser middleware will check if the cookie is signed or not